presto-accumulo/src/main/java/com/facebook/presto/accumulo/index/Indexer.java

```diff
@@ -0,0 +1,594 @@
+
+    /**
+     * Text version of METRICS_TABLE_ROWID
+     */
+    public static final Text METRICS_TABLE_ROWID_AS_TEXT = new Text(METRICS_TABLE_ROW_ID.array());
+
+    private static final byte[] EMPTY_BYTES = new byte[0];
+    private static final byte UNDERSCORE = '_';
+    private static final TypedValueCombiner.Encoder<Long> ENCODER =
+            new LongCombiner.StringEncoder();
+
+    private final AccumuloTable table;
+    private final BatchWriter indexWrtr;
+    private final BatchWriterConfig bwc;
+    private final Connector conn;
+    /**
+     * This map tracks cardinality for each column family and qualifier that is added to the index
+     */
+    private final Map<ByteBuffer, Map<ByteBuffer, AtomicLong>> metrics = new HashMap<>();
+
+    /**
+     * Mapping of column family to set of column qualifiers for all indexed Presto columns
+     */
+    private final Map<ByteBuffer, Set<ByteBuffer>> indexColumns = new HashMap<>();
+
+    /**
+     * Mapping of column family to column qualifier to Presto column type for all indexed Presto
+     * columns
+     */
+    private final Map<ByteBuffer, Map<ByteBuffer, Type>> indexColumnTypes = new HashMap<>();
+
+    /**
+     * Serializer class for the table
+     */
+    private final AccumuloRowSerializer serializer;
+
+    private Comparator<byte[]> byteArrayComparator = UnsignedBytes.lexicographicalComparator();
+    private byte[] firstRow = null;
+    private byte[] lastRow = null;
+
+    /**
+     * Creates a new instance of an {@link Indexer}
+     *
+     * @param conn Connector for interacting with Accumulo
+     * @param auths Authorizations for the particular user
+     * @param table Table metadata, see AccumuloClient#getTable
+     * @param bwc Config for the BatchWriter that will be writing index mutations
+     * @throws TableNotFoundException If the index and/or metric table do not exist
+     */
+    public Indexer(Connector conn, Authorizations auths, AccumuloTable table, BatchWriterConfig bwc)
+            throws TableNotFoundException
+    {
+        this.conn = conn;
+        this.table = table;
+        this.bwc = bwc;
+        this.serializer = table.getSerializerInstance();
+
+        // Create our batch writer
+        indexWrtr = conn.createBatchWriter(table.getIndexTableName(), bwc);
+
+        // Initialize metadata
+        table.getColumns().stream().forEach(col -> {
+            if (col.isIndexed()) {
+                // Wrap the column family and qualifier for this column
+                ByteBuffer cf = wrap(col.getFamily().getBytes());
+                ByteBuffer cq = wrap(col.getQualifier().getBytes());
+
+                // Get all qualifiers for this given column family, creating a new one if necessary
+                Set<ByteBuffer> qualifiers = indexColumns.get(cf);
+                if (qualifiers == null) {
+                    qualifiers = new HashSet<>();
+                    indexColumns.put(cf, qualifiers);
```

Comment:
Why not make `indexColumns` `ImmutableMap` with `ImmutableSet` inside? It looks like you're not changing the set outside c-tor?